### ISSUE TITLE
Implement globbing (expand *) on Windows

### ISSUE THREAD

**Issue by yhafri** [restic/restic#722]
## Output of `restic version`

```
C:\Users\Admin>restic version
restic 0.3.2
compiled with go1.7.4 on windows/amd64
```

Window *7* and *10*.

## Expected behavior

We should be able to specify wildcards when performing backups:

```
C:\Users\Admin>restic backup -f e:\Downloads\Test\*.*
```

## Actual behavior

Try with wildcard, you'll get an error message and it completes with an empty snapshot. Still no progress bar nor ETA.

On Linux, it works as expected.

## Steps to reproduce the behavior

```
C:\Users\Admin>restic backup -f e:\Downloads\Test\*.*
scan [e:\Downloads\Test\*.*]
error for \\?\e:\Downloads\Test\*.*: GetFileAttributesEx \\?\e:\Downloads\Test\*.*: The filename, directory name, or volume label syntax is incorrect.
scanned 0 directories, 0 files in 0:00
error walking dir *.*: Lstat: GetFileAttributesEx \\?\e:\Downloads\Test\*.*: The filename, directory name, or volume label syntax is incorrect.
[0:00]   0B/s  0B / 0B  0 / 0 items  1 errors  ETA 0:00
duration: 0:00, 0.00MiB/s
snapshot 79161d62 saved
 
C:\Users\Admin>restic snapshots
ID        Date                 Host        Tags        Directory
----------------------------------------------------------------------
6a9aa486  2017-01-10 03:45:20  THOR                    e:\Downloads\Test\
79161d62  2017-01-10 03:48:54  THOR                    e:\Downloads\Test\*.*
7e262feb  2017-01-10 03:49:28  THOR                    e:\Downloads\Test\
 
C:\Users\Admin>restic ls 79161d62
snapshot of [e:\Downloads\Test\*.*] at 2017-01-10 03:48:54.1638702 -0500 EST:
 ```

**Comment by fd0** [comment]
Hi, thanks for your report. This function is called "globbing" and on Unix systems, the user's shell is responsible for expanding the wildcard characters. So if you enter `restic backup f*`, the shell will expand the asterisk and runs restic with the expanded arguments, e.g. `restic backup foo foo2 fump foobar`, so restic doesn't even see the wildcards.

Apparently, this isn't done by `cmd.exe`, and all programs need to implement this by themselves. Here, restic gets the (literal) filename `e:\Downloads\Test\*.*`, which does not exist (there's no file called `*.*` in that directory).

Since this is very different from what restic does on Unix, I'm not convinced that restic needs this globbing function itself. Isn't there a way to get `cmd.exe` expand the arguments, like a shell does on Unix?

**Comment by yhafri** [comment]
Hmm ... I agree @fd. But the pattern ** \*.* ** is very common under Windows. Why other command line tools (ex. dir) support it and not _restic_ ?

Finally, there is 2 files inside _e:\Downloads\Test\_ 
Both have a name + an extension.

**Comment by fd0** [comment]
> Hmm ... I agree @fd0 but the pattern \*.* is very common under Windows. Why other command line tools (ex. dir, del) support it and not restic ?

I guess that all those commands implement globbing by themselves, maybe using a library.

>> Here, restic gets the (literal) filename e:\Downloads\Test*.*, which does not exist (there's no file called . in that directory).
> There is 2 files inside e:\Downloads\Test directory. Both have a name and an extension.

What I meant was: There is no file which name and extension each only consists of an asterisk character (`*`).

> If it's the case why restic proceeded with the backup, and even returned an ID (79161d62) then?

That's an excellent question, and I think this behavior is a bug. If nothing is backed up, restic should print an error message.

Yes, I'm using the `path` and `filepath` packages within restic. We can also use `filepath` to add globbing, but this function needs to be restricted to windows, because on Unix (as we already discussed) wildcards are expanded by the shell. So we'll end up with different behavior on Windows and Unix, and that's what I'm concerned about. The actual function should be easy to implement.

**Comment by yhafri** [comment]
Totally agree @fd0

Windows should be the only exception here because globbing works as expected elsewhere (Linux, MacOS).

I'm not sure we'll end up with different behaviour IMHO. It's just Windows and a fix is needed 😁

**Comment by notna123** [comment]
hmm... made an "--exclude-file" with stuff like "C:\$Recycle.Bin" and/or "C:\$Recycle.Bin\" and/or like "C:\$Recycle.Bin\*" but seems like this is also not working...

if it's related to this discussion here, the "solution" seems to be up to you guys... see "If someone else needs their program to expand command line parameters in a particular way, they should do so" [here ](https://github.com/golang/go/issues/3415)

**Comment by notna123** [comment]
pls note, seems like markdown swollows the \\'s

**Comment by fd0** [comment]
Thanks for the report. I've fixed the markdown, you can denote raw text with backticks ("`"), please review your comment.

This is a different issue, please open a new ticket. This issue is about expanding globbing in command-line parameters, you're talking about globbing patterns in an exclude file (which should work, so it looks like a bug).

**Comment by fd0** [comment]
@notna123 The problem here may be that the `c:\` isn't interpreted correctly. Could you try again with `\$Recycle.Bin`?

**Comment by notna123** [comment]
Still not working... seems like it works as expected if passed via "--exclude" in the comand line BUT non of the various paths I tried in a "--exclude-file" seems to work:

 


 

C:\>C:\bin\restic -r U:\Backup\restic\C_Drive\ backup --exclude "C:\$Recycle.Bin\*" C:\  <--- OK! (with \* at the end!)
enter password for repository:
scan [C:\]
error for \\?\C:\Config.Msi: open \\?\C:\Config.Msi: Access is denied.
Interrupt received, cleaning up
[0:04] 2925 directories, 37525 files, 2.297 GiB

C:\>C:\bin\restic -r U:\Backup\restic\C_Drive\ backup --exclude "C:\$Recycle.Bin\" C:\  <--- NOT OK! (only \ at the end!)
invalid argument "C:\\$Recycle.Bin\" C:\\" for --exclude: line 1, column 15: bare " in non-quoted-field

 

C:\>C:\bin\restic -r U:\Backup\restic\C_Drive\ backup --exclude "C:\$Recycle.Bin" C:\  <--- OK!
enter password for repository:
scan [C:\]
error for \\?\C:\Config.Msi: open \\?\C:\Config.Msi: Access is denied.
error for \\?\C:\Documents and Settings: open \\?\C:\Documents and Settings: Access is denied.
Interrupt received, cleaning up
[0:03] 3189 directories, 41631 files, 2.517 GiB

C:\>C:\bin\restic -r U:\Backup\restic\C_Drive\ backup --exclude C:\$Recycle.Bin C:\  <--- OK!
enter password for repository:
scan [C:\]
error for \\?\C:\Config.Msi: open \\?\C:\Config.Msi: Access is denied.
Interrupt received, cleaning up
[0:02] 2640 directories, 33941 files, 2.256 GiB
 

 

--->>> EXCLUDE-FILE not working at all <<<---


C:\>C:\bin\restic -r U:\Backup\restic\C_Drive\ backup --exclude-file C:\bin\restic_exclude_list_bin.txt C:\
enter password for repository:
scan [C:\]
error for \\?\C:\$Recycle.Bin\S-1-5-18: open \\?\C:\$Recycle.Bin\S-1-5-18: Access is denied.
error for \\?\C:\$Recycle.Bin\S-1-5-21-1278075657-2615590652-2769505701-1004: open \\?\C:\$Recycle.Bin\S-1-5-21-1278075657-2615590652-2769505701-1004: Access is denied.
error for \\?\C:\$Recycle.Bin\S-1-5-21-1278075657-2615590652-2769505701-1010: open \\?\C:\$Recycle.Bin\S-1-5-21-1278075657-2615590652-2769505701-1010: Access is denied.
error for \\?\C:\$Recycle.Bin\S-1-5-21-1278075657-2615590652-2769505701-1011: open \\?\C:\$Recycle.Bin\S-1-5-21-1278075657-2615590652-2769505701-1011: Access is denied.
error for \\?\C:\Config.Msi: open \\?\C:\Config.Msi: Access is denied.
Interrupt received, cleaning up
[0:02] 1549 directories, 10886 files, 2.015 GiB


C:\>more C:\bin\restic_exclude_list_bin.txt
"C:\$Recycle.Bin\*"
C:\$Recycle.Bin\*
C:\$Recycle.Bin\
C:\$Recycle.Bin
\$Recycle.Bin

C:\>
 

 


 

Gesendet: Dienstag, 11. April 2017 um 12:59 Uhr
Von: "Alexander Neumann" <notifications@github.com>
An: restic/restic <restic@noreply.github.com>
Cc: notna123 <notna@ist-einmalig.de>, Mention <mention@noreply.github.com>
Betreff: Re: [restic/restic] Implement globbing (expand *) on Windows (#722)


@notna123 The problem here may be that the c:\ isn't interpreted correctly. Could you try again with \$Recycle.Bin?

**Comment by fd0** [comment]
Oh, indeed, this looks like an issue with the "long paths" (prefixed with `\\?\`), this is some strange windows specific path thingy, which makes everything a lot more complicated...

**Comment by notna123** [comment]
exclude via command line seems to work fine, in opposite what ever I tried in an exclude-file never worked :|
 

Gesendet: Dienstag, 11. April 2017 um 07:38 Uhr
Von: "Alexander Neumann" <notifications@github.com>
An: restic/restic <restic@noreply.github.com>
Cc: notna123 <notna@ist-einmalig.de>, Comment <comment@noreply.github.com>
Betreff: Re: [restic/restic] Implement globbing (expand *) on Windows (#722)


Thanks for the report. I've fixed the markdown, you can denote raw text with backticks ("`"), please review your comment.

This is a different issue, please open a new ticket. This issue is about expanding globbing in command-line parameters, you're talking about globbing patterns in an exclude file (which should work, so it looks like a bug).

**Comment by bjoe2k4** [comment]
@AntonOks Excluding `$RECYCLE.BIN` via `--exclude-file` is now possible via `$$RECYCLE.BIN` after #1858. For all versions before that, another workaround can be found in #1857.

**Comment by filippobottega** [comment]
Hi,
  here [Is there any way to get the windows cmd shell to expand wildcard paths?](https://superuser.com/questions/460598/is-there-any-way-to-get-the-windows-cmd-shell-to-expand-wildcard-paths)  wullxz suggests to use Powershell to do the expansion in this way:
  `app.exe $(ls *.txt | % {$_.FullName})`

  Not tested but promising...

  Filippo

**Comment by filippobottega** [comment]
Hi,
  I have tested and it seems to work!

  Try this command using PowerShell and not cmd:

`restic_0.9.6_windows_amd64.exe backup -r <Repo> -p <Password> $(ls .git -Recurse -Directory -Hidden | % {$_.FullName} | Resolve-Path -Relative)`

  Filippo

**Comment by MichaelEischer** [comment]
Since restic 0.9.3 `--files-from` implements globbing (see #1892).

**Comment by MichaelEischer** [comment]
If you need glob expansion use `--files-from`. If you don't want glob expansion use `--files-from-verbatim` or pass the paths on the command line. Unfortunately there is no way from restic to guess whether a user expects paths on the command line to be interpreted as glob pattern or not. Blindly treating everything as a glob pattern, would break paths containing square brackets `[]`, so this isn't an option either.

Thus there is nothing left to be done here.
